go/libraries/doltcore/rebase/rebase.go

```diff
@@ -0,0 +1,458 @@
+
+	if err != nil {
+		return nil, err
+	}
+
+	cm, err := ddb.Resolve(ctx, cs)
+
+	commitHistory, otherParents, err := rewindCommitHistory(ctx, ddb, cm, oldTag)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(commitHistory) < 2 {
+		panic("need base commit and at least one change commit using oldTag")
+	}
+
+	parent := commitHistory[0]
+	rebasedParent := parent
+	commitHistory = commitHistory[1:]
+	var rebasedCommit *doltdb.Commit
+
+	for _, commit := range commitHistory {
+
+		root, err := commit.GetRootValue()
+
+		if err != nil {
+			return nil, err
+		}
+
+		parentRoot, err := parent.GetRootValue()
+
+		if err != nil {
+			return nil, err
+		}
+
+		rebasedParentRoot, err := rebasedParent.GetRootValue()
+
+		if err != nil {
+			return nil, err
+		}
+
+		rebasedRoot, err := replayCommitWithNewTag(ctx, root, parentRoot, rebasedParentRoot, oldTag, newTag)
+
+		if err != nil {
+			return nil, err
+		}
+
+		valueHash, err := ddb.WriteRootValue(ctx, rebasedRoot)
+
+		if err != nil {
+			return nil, err
+		}
+
+		oldMeta, err := commit.GetCommitMeta()
+
+		if err != nil {
+			return nil, err
+		}
+
+		ch, _ := commit.HashOf()
+		parents := append(otherParents[ch], rebasedParent)
+		rebasedCommit, err = ddb.CommitWithParentCommits(ctx, valueHash, dRef, parents, oldMeta)
+
+		if err != nil {
+			return nil, err
+		}
+
+		parent = commit
+		rebasedParent = rebasedCommit
+	}
+	return rebasedCommit, nil
+}
+
+func rewindCommitHistory(ctx context.Context, ddb *doltdb.DoltDB, c *doltdb.Commit, oldTag uint64) ([]*doltdb.Commit, parentMap, error) {
+
+	otherParents := make(parentMap)
+	var history []*doltdb.Commit
+	cur := c
+	for {
+		n, err := cur.NumParents()
+
+		if n == 0 {
+			// beginning of history & first commit uses oldTag
+			panic("can't rebase from the beginning of history")
+		}
+
+		if err != nil {
+			return nil, nil, err
+		}
+
+		h, _ := cur.HashOf()
+		otherParents[h] = []*doltdb.Commit{}
+		tagUsed := false
+		history = append(history, cur)
+
+		allParents, err := ddb.ResolveAllParents(ctx, cur)
+
+		if err != nil {
+			return nil, nil, err
+		}
+
+		for _, pc := range allParents {
+
+			tagUsedInParent, err := tagUsedInHistory(ctx, ddb, pc, oldTag)
+
+			if err != nil {
+				return nil, nil, err
+			}
+
+			if tagUsedInParent {
+				if tagUsed {
+					panic("tag can be used by at most one parent")
+				}
+				tagUsed = true
+				cur = pc
+			} else {
+				otherParents[h] = append(otherParents[h], pc)
+			}
+		}
+
+		if !tagUsed {
+			// reached first usage of oldTag
+			if len(otherParents[h]) != 1 {
+				panic("unreachable: if multiple parents exist, one must use the tag")
+			}
+			history = append(history, otherParents[h]...)
+			break
+		}
+	}
+
+	// reverse slice: oldest first
+	for i := 0; i < len(history)/2; i++ {
+		opp := len(history) - 1 - i
+		history[i], history[opp] = history[opp], history[i]
+	}
+
+	return history, otherParents, nil
+}
+
+// TODO: optimize this function with a visitedNode set
```

Comment:
I'm leaving the naive implementation for now. Storing the SuperSchema in the commit struct will allow this optimization in the future